tiledb/sm/c_api/tiledb.cc

```diff
@@ -3393,6 +3421,461 @@ int32_t tiledb_query_get_fragment_timestamp_range(
+}
+
+int32_t tiledb_subarray_get_range_from_name(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    const char* dim_name,
+    uint64_t range_idx,
+    const void** start,
+    const void** end,
+    const void** stride) {
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  if (SAVE_ERROR_CATCH(
+          ctx,
+          subarray->subarray_->get_range_from_name(
+              dim_name, range_idx, start, end, stride)))
+    return TILEDB_ERR;
+
+  return TILEDB_OK;
+}
+
+int32_t tiledb_subarray_get_range_var_size_from_name(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    const char* dim_name,
+    uint64_t range_idx,
+    uint64_t* start_size,
+    uint64_t* end_size) {
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  if (SAVE_ERROR_CATCH(
+          ctx,
+          subarray->subarray_->get_range_var_size_from_name(
+              dim_name, range_idx, start_size, end_size)))
+    return TILEDB_ERR;
+
+  return TILEDB_OK;
+}
+
+int32_t tiledb_subarray_get_range_var(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    uint32_t dim_idx,
+    uint64_t range_idx,
+    void* start,
+    void* end) {
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  if (SAVE_ERROR_CATCH(
+          ctx,
+          subarray->subarray_->get_range_var(dim_idx, range_idx, start, end)))
+    return TILEDB_ERR;
+
+  return TILEDB_OK;
+}
+
+int32_t tiledb_subarray_get_range_var_from_name(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    const char* dim_name,
+    uint64_t range_idx,
+    void* start,
+    void* end) {
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  if (SAVE_ERROR_CATCH(
+          ctx,
+          subarray->subarray_->get_range_var_from_name(
+              dim_name, range_idx, start, end)))
+    return TILEDB_ERR;
+
+  return TILEDB_OK;
+}
+
+int32_t tiledb_subarray_get_est_result_size(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    const char* name,
+    uint64_t* size) {
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+  if (SAVE_ERROR_CATCH(
+          ctx,
+          subarray->subarray_->get_est_result_size(
+              name, size, ctx->ctx_->storage_manager())))
+    return TILEDB_ERR;
+  return TILEDB_OK;
+}
+
+int32_t tiledb_subarray_get_est_result_size_var(
+    tiledb_ctx_t* ctx,
+    const tiledb_subarray_t* subarray,
+    const char* name,
+    uint64_t* size_off,
+    uint64_t* size_val) {
+  if (sanity_check(ctx) == TILEDB_ERR ||
+      sanity_check(ctx, subarray) == TILEDB_ERR)
+    return TILEDB_ERR;
+
+  if (SAVE_ERROR_CATCH(
+          ctx,
+          subarray->subarray_->get_est_result_size(
```

Comment:
This API doesn't include all the checks we have at the query level, let's rename the current one to _unsafe, then make get_est_result_size in subarray.cc that includes the same checks as in the query, which ends up calling the unsafe version.

Reply:
Same comment for tiledb_subarray_get_est_result_size_nullable and tiledb_subarray_get_est_result_size_var_nullable

Reply:
(note ..._result_size_var_..., just overloaded get_est_result_size_nullable() for that purpose)
As with earlier discussed ->get_est_result_size() routines, there currently is no _unsafe/_internal and these routines both contain significant checks.
Do you think these should be renamed _internal, or some other change?
(to me, _internal naming does not really seem appropriate for these, assuming we do continue with them being exposed via tiledb_subarray_...() external APIs.)

Reply:
This is not what I'm talking about. Let's look at tiledb_query_get_est_result_size_nullable. It calls Query::get_est_result_size_nullable, which does a lot of checks, then calls Reader::get_est_result_size_nullable, then Subarray::get_est_result_size_nullable.

Now tiledb_subarray_get_est_result_size_nullable, calls Subarray::get_est_result_size_nullable directly, without any of the checks. You are skipping a lot of checks that were done in that code path.

Reply:
Let's look at two different paths:
tiledb_query_get_est_result_size_nullable calls Query::get_est_result_size_nullable which does a lot of checking, then calls Subarray::get_est_result_size_nullable.

Now tiledb_subarray_get_est_result_size_nullable calls Subarray::get_est_result_size_nullable.

Do you see how your new api is missing the checks the old one has?

Reply:
Don't think _unsafe is appropriate naming for either of those Subarray overloaded methods...
(And think _internal conveys nothing useful since there is no corresponding non-internal method...)
Here's what I see...
I see that both are missing a single 'WRITE' check from their corresponding Query proxy 
(which perhaps *should* be in Subarray if functionality is to remain exposed from there, 
but current unit tests pass, maybe something missing there), 
but I also see that
Subarray::overload-A contains 5 if() checks, at least one of which Query does not contain
Subarray::overload-B contains  5 if() checks, some of which Query does not contain
which in my mind disqualifies it from being "_unsafe", 
different yes, but not "_unsafe".
???

Reply:
We already discussed _unsafe and decided to go with _internal for get_est_result_size. And I'm saying we should respect the same pattern we already established for get_est_result_size. The existing subarray API (get_est_result_size_nullable) gets an _internal added and we need to create a new get_est_result_size_nullable in the subarray that contains the same checks as Query::get_est_result_size_nullable that calls to Subarray::get_est_result_size_nullable_internal. The old Query::get_est_result_size_nullable would now call to Subarray::get_est_result_size_nullable_internal too.

Reply:
changes made